src/main/java/com/pinterest/secor/parser/QuboleClient.java

```diff
@@ -94,24 +96,28 @@ private int query(String query) throws IOException {
         return (Integer) response.get("id");
     }
 
-    private void waitForCompletion(int commandId) throws IOException, InterruptedException {
+    private int waitForCompletion(int commandId, long timeout) throws IOException, InterruptedException {
         URL url = new URL("https://api.qubole.com/api/v1.2/commands/" + commandId);
-        while (true) {
+        long endTime = System.currentTimeMillis() + timeout;
+
+        while (System.currentTimeMillis() < endTime) {
             Map response = makeRequest(url, null);
             if (response.get("status").equals("done")) {
-                return;
+                return 0;
             }
             System.out.println("waiting 3 seconds for results of query " + commandId +
                                ". Current status " + response.get("status"));
             Thread.sleep(3000);
         }
+
+        return 1; // qubole call failed to return within timeout
```

Comment:
Actually, it's not necessary to return anything. If an exception occurs, the task will exit and the next finalization run will take care of finalizing any partitions that did not succeed.